Dynamic notch changes

KarateBrot opened the pull request:
# Dynamic Notch Q
This PR reverts the changes to the dynamic notch Q value.

It turned out that using a constant width for the notches didn't give enough benefits. Using constant bandwidth makes sure notches don't get too narrow at low frequencies. Turns out we don't go so low as to get a real benefit from that. Also, at low frequencies wide notches are greatly contributing to propwash.

**We are looking at another way to improve the notch filter capabilities at low frequencies without latency costs.**

Reverting back to constant Q value gives the following benefits:

- notches get wider at higher freqs without contributing much to filter latency
- which means slightly less propwash at lower frequencies (tested and confirmed)
- and a bit more overall filter capabilities (tested and confirmed)
- people are used to Q value. They have a "feeling" for it
- less maintenance work for intercompatibility of different versions

## Tuning

The standard bandwidth value was pretty slim at 45Hz. That corresponds to a Q value of about 500. We can do that because the dynamic notch is pretty precise at double the frequency resolution thanks to the SDFT.

**Dynamic notch settings:**
```
set dyn_notch_count = 3
set dyn_notch_q = 300
```

**Dynamic notch settings with RPM filter:**
```
set dyn_notch_count = 1..2
set dyn_notch_q = 500
```

# Frequency estimation
In addition the frequency estimation algorithm got simplified. Everything still works just as expected.

I designed this algorithm independently but the same method (and more) is described in great detail in a paper from CERN called "[IMPROVING FFT FREQUENCY MEASUREMENT RESOLUTION  BY PARABOLIC AND GAUSSIAN INTERPOLATION](https://mgasior.web.cern.ch/pap/FFT_resol_note.pdf)".

## Fun(?) fact
Btw. for gaussian interpolation instead of parabolic interpolation we just need to calculate the natural log of y0, y1 and y2 before estimating meanBin (because a gauss curve looks like a parabola on log scale). Easy!

```c
float meanBin = peaks[p].bin;

const float y0 = sdftData[peaks[p].bin - 1];
const float y1 = sdftData[peaks[p].bin];
const float y2 = sdftData[peaks[p].bin + 1];

// Fit gauss curve instead of parabola
y0 = log_approx(y0);
y1 = log_approx(y1);
y2 = log_approx(y2);

// Estimate true peak position aka. meanBin (fit parabola y(x) over y0, y1 and y2, solve dy/dx=0 for x)
const float denom = 2.0f * (y0 - 2 * y1 + y2);
if (denom != 0.0f) {
    meanBin += (y0 - y2) / denom;
}
```

This improves the frequency estimation 4 times but the algorithm gets very inefficient, which is not worth it for "just" 4x less error.

Linked PRs:
https://github.com/betaflight/betaflight-configurator/pull/2504

KarateBrot commented:
@Asizon @klutvott123 @ctzsnooze 
Finally reverted the changes for the Q value. Everything else works just as expected. We can soon revert the changes for the configurator, LUA script and blackbox explorer as well.

ctzsnooze reviewed (approved):
Great!  Confident this will work really well

DusKing1 commented on src/main/blackbox/blackbox.c:
```diff
@@ -1422,7 +1422,7 @@ static bool blackboxWriteSysinfo(void)
         BLACKBOX_PRINT_HEADER_LINE("dyn_notch_max_hz", "%d",                gyroConfig()->dyn_notch_max_hz);
         BLACKBOX_PRINT_HEADER_LINE("dyn_notch_count", "%d",                 gyroConfig()->dyn_notch_count);
-        BLACKBOX_PRINT_HEADER_LINE("dyn_notch_bandwidth_hz", "%d",          gyroConfig()->dyn_notch_bandwidth_hz);
+        BLACKBOX_PRINT_HEADER_LINE("dyn_notch_q", "%d",                     gyroConfig()->dyn_notch_q);
```
white spaces instead of tabs?

KarateBrot replied:
Yes, white spaces according to coding style

DusKing1 commented on src/main/cli/settings.c:
```diff
@@ -662,7 +662,7 @@ const clivalue_t valueTable[] = {
 #if defined(USE_GYRO_DATA_ANALYSE)
     { "dyn_notch_count",            VAR_UINT8   | MASTER_VALUE, .config.minmaxUnsigned = { 1, DYN_NOTCH_COUNT_MAX }, PG_GYRO_CONFIG, offsetof(gyroConfig_t, dyn_notch_count) },
-    { "dyn_notch_bandwidth_hz",     VAR_UINT16  | MASTER_VALUE, .config.minmaxUnsigned = { 1, 1000 }, PG_GYRO_CONFIG, offsetof(gyroConfig_t, dyn_notch_bandwidth_hz) },
+    { "dyn_notch_q",                VAR_UINT16  | MASTER_VALUE, .config.minmaxUnsigned = { 1, 1000 }, PG_GYRO_CONFIG, offsetof(gyroConfig_t, dyn_notch_q) },
```
white spaces instead of tabs?

mikeller commented on src/main/blackbox/blackbox.c:
```diff
@@ -1422,7 +1422,7 @@ static bool blackboxWriteSysinfo(void)
         BLACKBOX_PRINT_HEADER_LINE("dyn_notch_max_hz", "%d",                gyroConfig()->dyn_notch_max_hz);
         BLACKBOX_PRINT_HEADER_LINE("dyn_notch_count", "%d",                 gyroConfig()->dyn_notch_count);
-        BLACKBOX_PRINT_HEADER_LINE("dyn_notch_bandwidth_hz", "%d",          gyroConfig()->dyn_notch_bandwidth_hz);
+        BLACKBOX_PRINT_HEADER_LINE("dyn_notch_q", "%d",                     gyroConfig()->dyn_notch_q);
```
This will also need coordination with the blackbox log viewer - have these changes already been factored into the work that @ctzsnooze is doing there?

ctzsnooze replied:
I’m setting blackbox log headers to display Q values, rather than bandwidth, for rpm and dynamic notch filters - so that’s all ok, I think.

KarateBrot commented:
@mikeller So I changed most integers smaller than machine word to machine word (32 bit). I also changed all the values used in calculations to float when it was possible. I tested the code, filtering still works. Actually minHz now gets rounded properly (it always rounded to ceiling before).

haslinghuis commented:
@KarateBrot 

> So I changed most integers smaller than machine word to machine word (32 bit).

Does this not make this more costly on flash size? You have replaced a lot of `uint8_t` to `int` here ?

KarateBrot commented:
@haslinghuis mikeller mentioned that using smaller sizes than machine word (in this case 32bit) will make the code less efficient because there needs to be range checks on them at runtime. And he was right. Changing it to 32bit brought down the filter loop from 20us to 15.5us on average. HUGE improvement in runtime. It brings up ITCM RAM usage a bit but flash size should still be the same.

Zuldan commented:
Any chance another 2 reviewers could have a look at this PR so it can be merged?

haslinghuis commented:
@KarateBrot @klutvott123 is there a PR for LUA / blackbox we can refer to for reference (added configurator link already) ?

mikeller reviewed (approved):
Not that it makes a big difference, but for things like loop counters that will never go negative, `unsigned` is a more appropriate machine word size type than `int`.

mikeller commented:
@KarateBrot:
> while the compiler might still change the smaller variables to 32bit at compile time, there also need to be range checks on them at runtime

There are no variables of less than 32 bit at assembly level - so this will always happen.

klutvott123 commented:
The PR for the lua scripts is here https://github.com/betaflight/betaflight-tx-lua-scripts/pull/383

KarateBrot commented:
Next time I can squash this down to 1 commit. I just thought leaving it as is (with 3 commits) would be easier for review. But thanks for the merge :)
